solar_consumer/save/save_data_platform.py

```diff
@@ -0,0 +1,173 @@
+            )
+            continue
+
+        # 7. Update location capacity based on the max capacity in this data
+        new_max_capacity_watts = int(gsp_data["capacity_mwp"].max() * 1_000_000)
+        max_capacity_watts_current = location.effective_capacity_watts
+        if new_max_capacity_watts > max_capacity_watts_current:
+            logger.info(
+                f"Updating location {location_uuid} capacity from {max_capacity_watts_current}W to {new_max_capacity_watts}W."
+            )
+
+            update_location_request = dp.UpdateLocationCapacityRequest(
+                location_uuid=location_uuid,
+                energy_source=dp.EnergySource.SOLAR,
+                valid_from_utc=datetime.now(tz=timezone.utc),
+                new_effective_capacity_watts=new_max_capacity_watts,
+            )
+            _ = await client.update_location_capacity(update_location_request)
+
+        # 8. Create new observations for the remaining data points
+        observation_values = []
+        for _, row in gsp_data.iterrows():
+            if row["capacity_mwp"] == 0:
+                continue  # Skip entries with zero capacity to avoid division by zero
+
+            value_fraction = row["solar_generation_kw"] / (row["capacity_mwp"] * 1000)
+            effective_capacity_watts = int(row["capacity_mwp"] * 1_000_000)
+            timestamp_utc = row["target_datetime_utc"].to_pydatetime()
+
+            oberservation_value = dp.CreateObservationsRequestValue(
+                timestamp_utc=timestamp_utc,
+                value_fraction=value_fraction,
+                effective_capacity_watts=effective_capacity_watts,
+            )
+            observation_values.append(oberservation_value)
+
+        if len(observation_values) > 0:
+
+            logger.debug(
+                f"Adding {len(observation_values)} new observation values for location {location_uuid} and observer {name}."
+            )
+
+            observation_request = dp.CreateObservationsRequest(
+                location_uuid=location_uuid,
+                energy_source=dp.EnergySource.SOLAR,
+                observer_name=name,
+                values=observation_values,
+            )
+            _ = await client.create_observations(observation_request)
+
+
+async def get_all_gsp_and_national_locations(
+    client: dp.DataPlatformDataServiceStub,
+) -> dict[int, dp.ListLocationsResponseLocationSummary]:
+    """Get all GSP and National locations for solar energy source"""
+
+    all_locations = {}
+
+    # National location
+    all_location_request = dp.ListLocationsRequest(
+        location_type_filter=dp.LocationType.NATION,
+        energy_source_filter=dp.EnergySource.SOLAR,
+    )
+    location_response = await client.list_locations(all_location_request)
+    all_uk_location = [loc for loc in location_response.locations if 'uk' in loc.location_name.lower()]
+    if len(all_uk_location) >= 1:
+        all_locations[0] = all_uk_location[0]
```

Comment:
What is this assumption now - for the case where we have more than one nation called UK? Surely it should error if there is more than one here?

Reply:
Ill put in a strong if and else and some errors